### PR TITLE
Selection history (back + forward button)

### PR DESCRIPTION
**Pull request by FliegendeWurst**
Continuation of https://git.key-project.org/key/key/-/merge_requests/629 (all review comments addressed)

### CODE REVIEW

File: key.ui/src/main/java/de/uka/ilkd/key/gui/MainWindow.java [hunk 1]
```diff
@@ -85,10 +85,6 @@ public final class MainWindow extends JFrame {
      * SequentView for the current goal
      */
     public final CurrentGoalView currentGoalView;
-    /**
-     * The menu for the SMT solver options
-     */
-    public final JMenu smtOptions = new JMenu("SMT Solvers...");
```
Comment (wadoon) [inline key.ui/src/main/java/de/uka/ilkd/key/gui/MainWindow.java:91]: I guess this field was unused. And why do we have public fields in `MainWindow`?

File: key.ui/src/main/java/de/uka/ilkd/key/gui/SelectionHistory.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ */
+public class SelectionHistory implements KeYSelectionListener, ProofDisposedListener {
+    /**
+     * The KeY mediator.
+     */
+    private final KeYMediator mediator;
+    /**
+     * Previously selected nodes by the user.
+     */
+    private final Deque<Node> selectedNodes = new ArrayDeque<>();
```
Comment (wadoon) [inline key.ui/src/main/java/de/uka/ilkd/key/gui/SelectionHistory.java:32]: This is dangerous! You created a memory leak. 

Keeping proof nodes alive, results into the case where proofs are kept alive (reachable via a pointer in node.), and when proofs are reachable, everything belonging to a proof is reachable. Reachable objects are not removed by GC. 

You should either use `WeakReference`s or try to store these information inside the user data of a proof.

**Review by wadoon — CHANGES_REQUESTED** [review]
Everything is fine. Only the memory leakage is an obstacles of acceptance.

File: key.ui/src/main/java/de/uka/ilkd/key/gui/SelectionHistory.java (hunk 2 above)
Comment (FliegendeWurst) [reply key.ui/src/main/java/de/uka/ilkd/key/gui/SelectionHistory.java]: Oh, thanks for the heads-up. I fixed it using `WeakReference`s.

File: key.ui/src/main/java/de/uka/ilkd/key/gui/MainWindow.java (hunk 1 above)
Comment (FliegendeWurst) [reply key.ui/src/main/java/de/uka/ilkd/key/gui/MainWindow.java:91]: Yes, the field was not used at all. I have made the other public fields private.

**Comment by WolframPfeifer** [comment]
@wadoon Merging is blocked at the moment. I think you have to approve that your requested changes have been applied ...

**Review by wadoon — APPROVED** [review]

**Pull request merged**